src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormattingPass.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+using TextSpan = Microsoft.CodeAnalysis.Text.TextSpan;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class CSharpFormattingPass : FormattingPassBase
+    {
+        private readonly ILogger _logger;
+
+        public CSharpFormattingPass(
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ILanguageServer server,
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<HtmlFormattingPass>();
+        }
+
+        // Run after the HTML formatter pass.
+        public override int Order => DefaultOrder - 5;
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result)
+        {
+            if (context.IsFormatOnType || result.Kind != RazorLanguageKind.Razor)
+            {
+                // We don't want to handle OnTypeFormatting here.
+                return result;
+            }
+
+            // Apply previous edits if any.
+            var originalText = context.CodeDocument.GetSourceText();
+            var changedText = originalText;
+            var changedContext = context;
+            if (result.Edits.Length > 0)
+            {
+                var changes = result.Edits.Select(e => e.AsTextChange(originalText)).ToArray();
+                changedText = changedText.WithChanges(changes);
```

Comment:
Given we're just applying these as-is. Does every pass do this? Wondering if this should be part of the orchestration

Reply:
I thought about passing in the changedContext to every successive pass (which would remove the need for all of these) but any pass after the first one would lose the guarantee that the `SourceText` they see will match the one on the client. So they can't use the client to format after the first pass. As of now, only the first pass (HtmlFormattingPass) uses the client to format so I can technically do this but I would rather do this cleanup when I know for sure I won't need to client after the first pass.

Reply:
ahhh I see